hoomd/mpcd/CollisionMethod.cc

```diff
@@ -48,15 +85,957 @@ void mpcd::CollisionMethod::collide(uint64_t timestep)
+    ArrayHandle<Scalar3> h_angmom_accum(m_angmom_accum, access_location::host, access_mode::read);
+
+    ArrayHandle<Scalar4> h_velocity(m_pdata->getVelocities(),
+                                    access_location::host,
+                                    access_mode::readwrite);
+    ArrayHandle<Scalar4> h_angmom(m_pdata->getAngularMomentumArray(),
+                                  access_location::host,
+                                  access_mode::readwrite);
+    ArrayHandle<Scalar4> h_orientation(m_pdata->getOrientationArray(),
+                                       access_location::host,
+                                       access_mode::read);
+    ArrayHandle<Scalar3> h_inertia(m_pdata->getMomentsOfInertiaArray(),
+                                   access_location::host,
+                                   access_mode::read);
+    ArrayHandle<unsigned int> h_rigid_center(m_rigid_bodies->getRigidCenters(),
+                                             access_location::host,
+                                             access_mode::read);
+    // add accumulated momentum to the central particle
+    unsigned int num_centers = m_rigid_bodies->getNLocal();
+    for (unsigned int idx = 0; idx < num_centers; ++idx)
+        {
+        // get index of central particle
+        const unsigned int central_idx = h_rigid_center.data[idx];
+
+        // get accumulated momentum for particle
+        const Scalar3 linmom_accum(h_linmom_accum.data[central_idx]);
+        const vec3<Scalar> angmom_accum(h_angmom_accum.data[central_idx]);
+
+        // compute and store new velocity
+        Scalar4 vel_mass = h_velocity.data[central_idx];
+        const Scalar mass = vel_mass.w;
+        if (mass > 0)
+            {
+            vel_mass.x += linmom_accum.x / mass;
+            vel_mass.y += linmom_accum.y / mass;
+            vel_mass.z += linmom_accum.z / mass;
+            h_velocity.data[central_idx] = vel_mass;
+            }
+
+        // compute new angular momentum
+        quat<Scalar> angmom(h_angmom.data[central_idx]);
+        const quat<Scalar> orientation(h_orientation.data[central_idx]);
+
+        // convert angular momentum to quaternion and update
+        const vec3<Scalar> inertia(h_inertia.data[central_idx]);
+        vec3<Scalar> angmom_change_body = rotate(conj(orientation), angmom_accum);
+        if (inertia.x == Scalar(0))
+            {
+            angmom_change_body.x = Scalar(0);
+            }
+
+        if (inertia.y == Scalar(0))
+            {
+            angmom_change_body.y = Scalar(0);
+            }
+
+        if (inertia.z == Scalar(0))
+            {
+            angmom_change_body.z = Scalar(0);
+            }
+        angmom += Scalar(2.0) * orientation * quat(0.0, angmom_change_body);
+
+        // save update
+        h_angmom.data[central_idx] = quat_to_scalar4(angmom);
+        }
+    }
+
+#ifdef ENABLE_HIP
+//! Create autotuners
+void mpcd::CollisionMethod::checkRigidAutotuners()
```

Comment:
This method has a few problems as written.
* You already created the autotuners in the constructor, so I don't think you need to do that again here. They can be removed.
* You can't swap the vector of autotuners because this may drop tuners a child has implemented.

Instead, you need to loop over the vector and only insert if they are not found. One way you might do this programatically is by making a vector like you did, then running a loop over them and checking with `std::find` if they need be inserted.

You also need an else branch that will loop over the vector of tuners and *removes* them if they are in the list but not needed. That will work similarly as above.

Reply:
I changed it to do that now.